### Fitting Logistic Regression

In this first notebook, you will be fitting a logistic regression model to a dataset where we would like to predict if a transaction is fraud or not.

To get started let's read in the libraries and take a quick look at the dataset.

In [1]:
import numpy as np
import pandas as pd
import statsmodels.api as sm

df = pd.read_csv('./fraud_dataset.csv')
df.head()

/opt/conda/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


transaction_id   duration      day  fraud
0           28891  21.302600  weekend  False
1           61629  22.932765  weekend  False
2           53707  32.694992  weekday  False
3           47812  32.784252  weekend  False
4           43455  17.756828  weekend  False

`1.` As you can see, there are two columns that need to be changed to dummy variables.  Replace each of the current columns to the dummy version.  Use the 1 for `weekday` and `True`, and 0 otherwise.  Use the first quiz to answer a few questions about the dataset.

In [2]:
# Create a dummy column for 'day', choose 'weekday' column to show
df['weekday'] = pd.get_dummies(df['day'])['weekday']

# Create 2 dummy columns for 'fraud' and 'not_fraud'
df[['not_fraud', 'fraud']] = pd.get_dummies(df['fraud'])

df = df.drop('not_fraud', axis=1)
df.head()

transaction_id   duration      day  fraud  weekday
0           28891  21.302600  weekend      0        0
1           61629  22.932765  weekend      0        0
2           53707  32.694992  weekday      0        1
3           47812  32.784252  weekend      0        0
4           43455  17.756828  weekend      0        0

In [3]:
# The proportion of fraudulent transactions
df['fraud'].mean()

0.012168770612987604

In [4]:
# The average duration for fraudulent transaction
df.query('fraud == 1')['duration'].mean()

4.6242473706156568

In [5]:
# The proportion of weekday transactions
df['weekday'].mean()

0.34527465029000343

In [6]:
# The average duration for non-fraudulent transactions
df.query('fraud == 0')['duration'].mean()

30.013583132522555

`2.` Now that you have dummy variables, fit a logistic regression model to predict if a transaction is fraud using both day and duration.  Don't forget an intercept!  Use the second quiz below to assure you fit the model correctly. Also remember to use the `.summary2() method to get your summary results.

In [7]:
# instantiate logistic regression model
df['intercept']= 1
logit_mod =sm.Logit(df['fraud'],df[['intercept', 'weekday', 'duration']])

# fit the model
results = logit_mod.fit()

# get summary statistics
results.summary2() 

/opt/conda/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py:1214: RuntimeWarning: overflow encountered in exp
  return 1/(1+np.exp(-X))
/opt/conda/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py:1264: RuntimeWarning: divide by zero encountered in log
  return np.sum(np.log(self.cdf(q*np.dot(X,params))))


Optimization terminated successfully.
         Current function value: inf
         Iterations 16


/opt/conda/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py:2397: RuntimeWarning: invalid value encountered in double_scalars
  return 1 - self.llf/self.llnull


<class 'statsmodels.iolib.summary2.Summary'>
"""
                        Results: Logit
==============================================================
Model:              Logit            No. Iterations:   16.0000
Dependent Variable: fraud            Pseudo R-squared: nan    
Date:               2023-07-10 16:10 AIC:              inf    
No. Observations:   8793             BIC:              inf    
Df Model:           2                Log-Likelihood:   -inf   
Df Residuals:       8790             LL-Null:          -inf   
Converged:          1.0000           Scale:            1.0000 
---------------------------------------------------------------
            Coef.   Std.Err.     z     P>|z|    [0.025   0.975]
---------------------------------------------------------------
intercept   9.8709    1.9438   5.0783  0.0000   6.0613  13.6806
weekday     2.5465    0.9043   2.8160  0.0049   0.7741   4.3188
duration   -1.4637    0.2905  -5.0389  0.0000  -2.0331  -0.8944
==============================================================

"""

In [8]:
np.exp(-1.4637), np.exp(2.5465)

(0.23137858821179411, 12.762357271496972)

In [9]:
1/np.exp(-1.4637)

4.3219210892783329

* On weekdays, the chance of fraud is 12.76 times more likely than on weekends, holding duration constant
* For each minute less spent on the transaction, the chance of fraud is 4.32 times more likely, holding the day of week constant